README.md
# nRF DFU Flashing Tool

[![crates.io](https://img.shields.io/crates/v/nrfdfu.svg)](https://crates.io/crates/nrfdfu)

`nrfdfu-rs` is an implementation of the protocol used by the bootloader on the nRF family of microcontrollers.
It can be used to flash an ELF firmware file onto devices such as the [nRF52840 Dongle].

This tool was written to replace [`pc-nrfutil`] in our trainings, so it implements a subset of the features found there.

[nRF52840 Dongle]: https://www.nordicsemi.com/Software-and-tools/Development-Kits/nRF52840-Dongle
[`pc-nrfutil`]: https://github.com/NordicSemiconductor/pc-nrfutil

## Usage

Run the following command to install the `nrfdfu` executable on your system:

```shell
$ cargo install nrfdfu
```

The tool is designed to be passed an ELF file as follows:

```
$ nrfdfu path/to/firmware.elf
```

This allows using it as a Cargo runner to automatically flash your Rust firmware during `cargo run`.
Place the following in `.cargo/config.toml` to use `nrfdfu` as the Cargo runner:

```toml
[target.'cfg(all(target_arch = "arm", target_os = "none"))']
runner = "nrfdfu"
```


CHANGELOG.md
# Changelog

## 0.1.3

- Relax firmware address check to support development boards other than the nRF-Dongle

## 0.1.2

- Improve error messages
- Change crc32 dependency

## 0.1.1

- Fixed timeouts on Windows

## 0.1.0

- Initial release


src/elf.rs
use std::iter;

use object::{
    elf::{FileHeader32, PT_LOAD},
    read::elf::{FileHeader, ProgramHeader, SectionHeader},
    Endianness, FileKind,
};

use crate::Result;

pub fn read_elf_image(elf: &[u8]) -> Result<Vec<u8>> {
    struct Chunk<'a> {
        flash_addr: u32,
        data: &'a [u8],
    }

    let file_kind = object::FileKind::parse(elf)
        .map_err(|e| format!("failed to parse firmware as ELF file: {}", e))?;

    if !matches!(file_kind, FileKind::Elf32) {
        return Err(format!(
            "firmware file has unsupported format {:?} (only 32-bit ELF files are supported)",
            file_kind
        )
        .into());
    }

    // Collect the to-be-flashed chunks.
    let mut chunks = Vec::new();

    let header = FileHeader32::<Endianness>::parse(elf)?;
    let endian = header.endian()?;
    let sections = header.section_headers(endian, elf)?;
    let strings = header.section_strings(endian, elf, sections)?;
    for (i, program) in header.program_headers(endian, elf)?.iter().enumerate() {
        let data = program
            .data(endian, elf)
            .map_err(|()| format!("failed to load segment data (corrupt ELF?)"))?;
        let p_type = program.p_type(endian);

        if !data.is_empty() && p_type == PT_LOAD {
            let (prog_offset, prog_size) = program.file_range(endian);

            // Note: `skip(1)` to skip the SHN_UNDEF at index 0
            let contains_section = sections.iter().skip(1).enumerate().any(|(sidx, section)| {
                let (sec_offset, sec_size) = match section.file_range(endian) {
                    Some(range) => range,
                    None => return false,
                };

                let contained =
                    sec_offset >= prog_offset && sec_offset + sec_size <= prog_offset + prog_size;
                if contained {
                    let name = String::from_utf8_lossy(section.name(endian, strings).unwrap());
                    log::debug!("phdr #{} contains section #{} {}", i, sidx, name);
                }
                contained
            });

            if contains_section {
                chunks.push(Chunk {
                    flash_addr: program.p_paddr(endian),
                    data,
                });
            }
        }
    }

    chunks.sort_by_key(|chunk| chunk.flash_addr);
    for ch in chunks.windows(2) {
        if ch[1].flash_addr < ch[0].flash_addr + ch[0].data.len() as u32 {
            return Err(format!("overlapping chunks at {:#x}", ch[1].flash_addr).into());
        }
    }

    if chunks.is_empty() {
        return Err(format!(
            "no loadable program segments found; ensure that the linker is \
            invoked correctly (passing the linker script)"
        )
        .into());
    }

    let mut image = Vec::new();
    let mut addr = chunks[0].flash_addr;
    log::debug!("firmware starts at {:#x}", addr);
    if addr < 0x1000 {
        return Err(format!(
            "firmware starts at address {:#x}, expected an address equal or higher than 0x1000 to \
             avoid a collision with the bootloader", addr).into());
    }

    for chunk in &chunks {
        if chunk.flash_addr < addr {
            return Err(format!(
                "overlapping program segments at 0x{:08x} (corrupt ELF?)",
                chunk.flash_addr
            )
            .into());
        }

        // Fill gaps between chunks with 0 bytes.
        let gap = chunk.flash_addr - addr;
        image.extend(iter::repeat(0).take(gap as usize));
        if gap > 0 {
            log::debug!("0x{:08x}-0x{:08x} (gap)", addr, addr + gap - 1);
        }
        addr += gap;

        image.extend(chunk.data);

        log::debug!(
            "0x{:08x}-0x{:08x}",
            chunk.flash_addr,
            chunk.flash_addr as usize + chunk.data.len() - 1
        );
        addr += chunk.data.len() as u32;
    }

    Ok(image)
}


src/main.rs
use log::LevelFilter;
use serialport::{available_ports, SerialPort};
use std::convert::TryInto;
use std::hash::Hasher;
use std::time::Duration;
use std::{error::Error, fs};

#[macro_use]
mod macros;
mod elf;
mod init_packet;
mod messages;
mod slip;

use messages::*;

type Result<T> = std::result::Result<T, Box<dyn Error>>;

const USB_VID: u16 = 0x1915;
const USB_PID: u16 = 0x521f;

/// Bootloader protocol version we support.
const PROTOCOL_VERSION: u8 = 1;

fn main() {
    match run() {
        Ok(()) => {}
        Err(e) => {
            eprintln!("error: {}", e);
            std::process::exit(1);
        }
    }
}

fn run() -> Result<()> {
    // We show info and higher levels by default, but allow overriding this via `RUST_LOG`.
    env_logger::builder()
        .filter_level(LevelFilter::Info)
        .parse_default_env()
        .init();

    let elf_path = std::env::args_os()
        .skip(1)
        .next()
        .ok_or_else(|| format!("missing argument (expected path to ELF file)"))?;
    let elf = fs::read(&elf_path)
        .map_err(|e| format!("couldn't read `{}`: {}", elf_path.to_string_lossy(), e))?;
    let mut image = elf::read_elf_image(&elf)?;

    let matching_ports: Vec<_> = available_ports()?
        .into_iter()
        .filter(|port| match &port.port_type {
            serialport::SerialPortType::UsbPort(usb) => usb.vid == USB_VID && usb.pid == USB_PID,
            _ => false,
        })
        .collect();

    let mut port = match matching_ports.len() {
        0 => {
            return Err(format!(
                "no matching USB serial device found.\n       Remember to put the \
                                 device in bootloader mode by pressing the reset button!"
            )
            .into())
        }
        1 => {
            let port = &matching_ports[0].port_name;
            log::debug!("opening {} (type {:?})", port, matching_ports[0].port_type);
            serialport::new(port, 115200)
                .timeout(Duration::from_millis(1000))
                .open()?
        }
        _ => return Err(format!("multiple matching USB serial devices found").into()),
    };

    // On Windows, this is required, otherwise communication fails with timeouts
    // (or just hangs forever).
    port.write_data_terminal_ready(true)?;

    let mut conn = BootloaderConnection::new(port)?;

    // Disable receipt notification. USB is a reliable transport.
    conn.set_receipt_notification(0)?;

    let obj_select = conn.select_object_command();
    log::debug!("select object response: {:?}", obj_select);

    let version = conn.fetch_protocol_version()?;
    log::debug!("protocol version: {}", version);

    let hw_version = conn.fetch_hardware_version()?;
    log::debug!("hardware version: {:?}", hw_version);

    // The firmware image must be padded with 0xFF to be a multiple of 4 Bytes. To our knowledge,
    // this is undocumented.
    while image.len() % 4 != 0 {
        image.push(0xff);
    }

    let init_packet = init_packet::build_init_packet(&image);
    conn.send_init_packet(&init_packet)?;
    conn.send_firmware(&image)?;

    Ok(())
}

struct BootloaderConnection {
    serial: Box<dyn SerialPort>,
    buf: Vec<u8>,
    mtu: u16,
}

impl BootloaderConnection {
    fn new(serial: Box<dyn SerialPort>) -> Result<Self> {
        let mut this = Self {
            serial,
            buf: Vec::new(),
            mtu: 0,
        };

        // We must check the protocol version before doing anything else, since any other command
        // might change if the version changes.
        let proto_version = this.fetch_protocol_version()?;
        if proto_version != PROTOCOL_VERSION {
            return Err(format!(
                "device reports protocol version {}, we only support {}",
                proto_version, PROTOCOL_VERSION
            )
            .into());
        }

        let mtu = this.fetch_mtu()?;
        log::debug!("MTU = {} Bytes", mtu);
        this.mtu = mtu;
        Ok(this)
    }

    /// send `req` and do not fetch any response
    fn request<R: Request>(&mut self, req: R) -> Result<()> {
        let mut buf = vec![R::OPCODE as u8];
        req.write_payload(&mut buf)?;
        log::trace!("--> {:?}", buf);

        // Go through an intermediate buffer to avoid writing every byte individually.
        self.buf.clear();
        slip::encode_frame(&buf, &mut self.buf)?;
        self.serial
            .write_all(&self.buf)
            .map_err(|e| format!("error while writing to serial port: {}", e))?;
        self.serial.flush()?;

        Ok(())
    }

    /// send `req` and expect a response.
    /// aborts if no response is received within timeout window.
    fn request_response<R: Request>(&mut self, req: R) -> Result<R::Response> {
        self.request(req)?;

        self.buf.clear();
        slip::decode_frame(&mut self.serial, &mut self.buf)
            .map_err(|e| format!("error while reading from serial port: {}", e))?;
        log::trace!("<-- {:?}", self.buf);

        messages::parse_response::<R>(&self.buf)
    }

    fn fetch_protocol_version(&mut self) -> Result<u8> {
        let response = self.request_response(ProtocolVersionRequest);
        match response {
            Ok(version_response) => Ok(version_response.version),
            Err(e) => Err(e),
        }
    }

    fn fetch_hardware_version(&mut self) -> Result<HardwareVersionResponse> {
        self.request_response(HardwareVersionRequest)
    }

    /// Sends the `.dat` file that's zipped into our firmware DFU .zip(?)
    /// modeled after `pc-nrfutil`s `dfu_transport_serial::send_init_packet()`
    fn send_init_packet(&mut self, data: &[u8]) -> Result<()> {
        log::info!("Sending init packet...");
        let select_response = self.select_object_command()?;
        log::debug!("Object selected: {:?}", select_response);

        let data_size = data.len() as u32;

        log::debug!("Creating Command...");
        self.create_command_object(data_size)?;
        log::debug!("Command created");

        log::debug!("Streaming Data: len: {}", data_size);
        self.stream_object_data(data)?;

        let received_crc = self.get_crc()?.crc;
        self.check_crc(data, received_crc, 0)?;

        self.execute()?;

        Ok(())
    }

    /// Sends the firmware image at `bin_path`.
    /// This is done in chunks to avoid exceeding our MTU  and involves periodic CRC checks.
    fn send_firmware(&mut self, image: &[u8]) -> Result<()> {
        log::info!("Sending firmware image of size {}...", image.len());

        log::debug!("Selecting Object: type Data");
        let select_response = self.select_object_data()?;
        log::debug!("Object selected: {:?}", select_response);

        let max_size = select_response.max_size;
        let mut prev_chunk_crc: u32 = 0;

        for chunk in image.chunks(max_size.try_into().unwrap()) {
            let curr_chunk_sz: u32 = chunk.len().try_into().unwrap();
            self.create_data_object(curr_chunk_sz)?;
            log::debug!("Streaming Data: len: {}", curr_chunk_sz);

            self.stream_object_data(chunk)?;

            let received_crc = self.get_crc()?;
            log::debug!("crc response: {:?}", received_crc);
            prev_chunk_crc = self.check_crc(chunk, received_crc.crc, prev_chunk_crc)?;

            self.execute()?;
        }

        log::info!("Done.");
        Ok(())
    }

    fn check_crc(&self, data: &[u8], received_crc: u32, initial: u32) -> Result<u32> {
        let mut digest = crc32fast::Hasher::new_with_initial(initial);
        digest.write(data);
        let expected_crc = digest.finalize();

        if expected_crc == received_crc {
            log::debug!("crc passed.");
            Ok(expected_crc)
        } else {
            let err_msg = format!(
                "crc failed: expected {} - received {}",
                expected_crc, received_crc
            );
            log::debug!("{}", err_msg);
            Err(err_msg.into())
        }
    }

    /// Sends a
    /// Request Type: `Select`
    /// Parameters:   `Object type = Command`
    fn select_object_command(&mut self) -> Result<SelectResponse> {
        self.request_response(SelectRequest(ObjectType::Command))
    }

    /// Sends a
    /// Request Type: `Select`
    /// Parameters:   `Object type = Data`
    fn select_object_data(&mut self) -> Result<SelectResponse> {
        self.request_response(SelectRequest(ObjectType::Data))
    }

    /// Sends a
    /// Request Type: `Create`
    /// Parameters:   `Object type = Command`
    ///               `size`
    fn create_command_object(&mut self, size: u32) -> Result<()> {
        self.request_response(CreateObjectRequest {
            obj_type: ObjectType::Command,
            size,
        })?;
        Ok(())
    }

    /// Sends a
    /// Request Type: `Create`
    /// Parameters:   `Object type = Data`
    ///               `size`
    fn create_data_object(&mut self, size: u32) -> Result<()> {
        // Note: Data objects cannot be created if no init packet has been sent. This results in an
        // `OperationNotPermitted` error.
        self.request_response(CreateObjectRequest {
            obj_type: ObjectType::Data,
            size,
        })?;
        Ok(())
    }

    fn set_receipt_notification(&mut self, every_n_packets: u16) -> Result<()> {
        self.request_response(SetPrnRequest(every_n_packets))?;
        Ok(())
    }

    fn fetch_mtu(&mut self) -> Result<u16> {
        Ok(self.request_response(GetMtuRequest)?.0)
    }

    fn stream_object_data(&mut self, data: &[u8]) -> Result<()> {
        // On the wire, the write request contains the opcode byte, and is then SLIP-encoded,
        // potentially doubling the size, and adding a frame terminator, so the chunk size has
        // to be smaller than the MTU.
        let max_chunk_size = usize::from((self.mtu - 1) / 2 - 1);

        for chunk in data.chunks(max_chunk_size) {
            // TODO: this also needs to take into account the receipt response. In our case we turn
            // it off, so there's nothing to do here.
            self.request(WriteRequest {
                request_payload: chunk,
            })?;
        }

        Ok(())
    }

    fn get_crc(&mut self) -> Result<CrcResponse> {
        self.request_response(CrcRequest)
    }

    // tell the target to execute whatever request setup we sent them before
    fn execute(&mut self) -> Result<ExecuteResponse> {
        self.request_response(ExecuteRequest)
    }
}


src/slip.rs
use std::io::{self, Read, Write};

const END: u8 = 0xC0;
const ESC: u8 = 0xDB;
const ESC_END: u8 = 0xDC;
const ESC_ESC: u8 = 0xDD;

pub fn encode_frame(buf: &[u8], mut writer: impl Write) -> io::Result<()> {
    for &byte in buf {
        match byte {
            END => writer.write_all(&[ESC, ESC_END])?,
            ESC => writer.write_all(&[ESC, ESC_ESC])?,
            _ => writer.write_all(&[byte])?,
        }
    }

    writer.write_all(&[END])?;

    Ok(())
}

pub fn decode_frame(reader: impl Read, buf: &mut Vec<u8>) -> io::Result<()> {
    let mut bytes = reader.bytes();
    loop {
        let encoded_byte = match bytes.next() {
            Some(byte) => byte,
            None => return Err(io::ErrorKind::UnexpectedEof.into()),
        };

        let decoded_byte = match encoded_byte? {
            ESC => match bytes.next() {
                None => return Err(io::ErrorKind::UnexpectedEof.into()),
                Some(Ok(ESC_ESC)) => ESC,
                Some(Ok(ESC_END)) => END,
                Some(Ok(invalid)) => {
                    return Err(io::Error::new(
                        io::ErrorKind::InvalidData,
                        format!("invalid byte following ESC: 0x{:02x}", invalid),
                    ))
                }
                Some(Err(e)) => return Err(e),
            },
            END => return Ok(()),
            other => other,
        };

        buf.push(decoded_byte);
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn encode(buf: &[u8]) -> Vec<u8> {
        let mut out = Vec::new();
        encode_frame(buf, &mut out).unwrap();
        out
    }

    fn decode(mut buf: &[u8]) -> Vec<u8> {
        let mut out = Vec::new();
        decode_frame(&mut buf, &mut out).unwrap();
        out
    }

    #[test]
    fn test_encode() {
        assert_eq!(encode(&[0]), vec![0, END]);
        assert_eq!(encode(&[END, 9]), vec![ESC, ESC_END, 9, END]);
        assert_eq!(
            encode(&[0, END, ESC, 1]),
            vec![0, ESC, ESC_END, ESC, ESC_ESC, 1, END]
        );
    }

    #[test]
    fn test_decode() {
        assert_eq!(decode(&[0, END]), vec![0]);
        assert_eq!(decode(&[ESC, ESC_END, 9, END]), vec![END, 9]);
        assert_eq!(
            decode(&[0, ESC, ESC_END, ESC, ESC_ESC, 1, END]),
            vec![0, END, ESC, 1]
        );
    }
}


src/macros.rs
/// Creates an enum that can be converted from and to a primitive type.
///
/// `Enum::from_primitive` will return an `Option<Enum>` corresponding to the matching variant.
///
/// This was copied almost verbatim from [smoltcp] and has been adapted to provide just conversion.
///
/// [smoltcp]: https://github.com/m-labs/smoltcp/blob/cd893e6ab60f094d684b37be7bc013bf79f0459d/src/macros.rs
macro_rules! primitive_enum {
    (
        $( #[$enum_attr:meta] )*
        $v:vis enum $name:ident($ty:ty) {
            $(
              $( #[$variant_attr:meta] )*
              $variant:ident = $value:expr
            ),* $(,)?
        }
    ) => {
        $( #[$enum_attr] )*
        $v enum $name {
            $(
              $( #[$variant_attr] )*
              $variant,
            )*
        }

        impl $name {
            $v fn from_primitive(prim: $ty) -> Option<Self> {
                match prim {
                    $( $value => Some($name::$variant), )*
                    _ => None,
                }
            }
        }

        impl ::core::convert::From<$name> for $ty {
            fn from(value: $name) -> Self {
                match value {
                    $( $name::$variant => $value, )*
                }
            }
        }
    }
}


src/dfu_package.rs
use std::iter;

use object::{
    elf::{FileHeader32, PT_LOAD},
    read::elf::{FileHeader, ProgramHeader},
    Bytes, Endianness, FileKind,
};

use crate::{init_packet::build_init_packet, Result};

pub struct DfuPackage {
    pub init_packet: Vec<u8>,
    pub image: Vec<u8>,
}

impl DfuPackage {
    pub fn from_elf(elf: &[u8]) -> Result<Self> {
        struct Chunk<'a> {
            flash_addr: u32,
            data: &'a [u8],
        }

        let file_kind = object::FileKind::parse(elf)
            .map_err(|e| format!("failed to parse firmware as ELF file: {}", e))?;
        if !matches!(file_kind, FileKind::Elf32) {
            return Err(format!(
                "firmware file has unsupported format {:?} (only 32-bit ELF files are supported)",
                file_kind
            )
            .into());
        }

        // Collect the to-be-flashed chunks.
        let mut chunks = Vec::new();

        let header = FileHeader32::<Endianness>::parse(Bytes(elf))?;
        let endian = header.endian()?;
        for program in header.program_headers(endian, Bytes(elf))? {
            let data = program
                .data(endian, Bytes(elf))
                .map_err(|()| format!("failed to load segment data (corrupt ELF?)"))?;
            let p_type = program.p_type(endian);

            if !data.is_empty() && p_type == PT_LOAD {
                chunks.push(Chunk {
                    flash_addr: program.p_paddr(endian),
                    data: data.0,
                });
            }
        }

        chunks.sort_by_key(|chunk| chunk.flash_addr);

        if chunks.is_empty() {
            return Err(format!("no loadable program segments found").into());
        }

        let mut image = Vec::new();
        let mut addr = chunks[0].flash_addr;
        eprintln!("firmware starts at {:#x}", addr);
        for chunk in &chunks {
            if chunk.flash_addr < addr {
                return Err(format!(
                    "overlapping program segments at 0x{:08x} (corrupt ELF?)",
                    chunk.flash_addr
                )
                .into());
            }

            // Fill gaps between chunks with erased 0xFF bytes.
            let gap = chunk.flash_addr - addr;
            image.extend(iter::once(0xFF).take(gap as usize));
            if gap > 0 {
                eprintln!("0x{:08x}-0x{:08x} (gap)", addr, chunk.flash_addr - 1);
            }

            image.extend(chunk.data);

            eprintln!(
                "0x{:08x}-0x{:08x}",
                chunk.flash_addr,
                chunk.flash_addr as usize + chunk.data.len()
            );
            addr += chunk.data.len() as u32;
        }

        Ok(Self {
            init_packet: build_init_packet(&image),
            image,
        })
    }
}


src/init_packet.rs
//! Implements serialization of "init packets", which contain firmware metadata and precede the
//! actual firmware upload.
//!
//! The full init command format is defined [here][init].
//!
//! [init]: https://github.com/tmael/nRF5_SDK/blob/master/components/libraries/bootloader/dfu/dfu-cc.proto

use rohs::WireType;
use sha2::{Digest, Sha256};

/// Tiny protobuf writer shim, free of `Pb`.
mod rohs {
    use std::mem;

    pub enum WireType {
        Varint = 0,
        LengthDelimited = 2,
    }

    pub trait Value {
        const TYPE: WireType;
        fn write(&self, writer: &mut MessageWriter);
    }

    pub trait Message {
        fn write(&self, writer: &mut MessageWriter);
    }

    impl<M: Message> Value for M {
        const TYPE: WireType = WireType::LengthDelimited;

        fn write(&self, writer: &mut MessageWriter) {
            // Nested messages are prefixed with their encoded length.

            let prev_buf = mem::replace(&mut writer.buf, Vec::new());
            <M as Message>::write(self, writer);
            let message = mem::replace(&mut writer.buf, prev_buf);
            writer.write_varint(message.len() as u64);
            writer.buf.extend(message);
        }
    }

    impl Value for bool {
        const TYPE: WireType = WireType::Varint;

        fn write(&self, writer: &mut MessageWriter) {
            writer.write_varint(*self as _);
        }
    }

    impl Value for u32 {
        const TYPE: WireType = WireType::Varint;

        fn write(&self, writer: &mut MessageWriter) {
            writer.write_varint(*self as _);
        }
    }

    impl Value for [u8] {
        const TYPE: WireType = WireType::LengthDelimited;

        fn write(&self, writer: &mut MessageWriter) {
            writer.write_varint(self.len() as _);
            writer.buf.extend(self.iter().copied());
        }
    }

    pub struct MessageWriter {
        buf: Vec<u8>,
    }

    impl MessageWriter {
        pub fn new() -> Self {
            Self { buf: Vec::new() }
        }

        pub fn write_field<V: Value + ?Sized>(&mut self, name: &str, field_number: u32, value: &V) {
            // `name` is only for documentation purposes
            let _ = name;

            let wire_type = V::TYPE as u64;
            let key = (u64::from(field_number) << 3) | wire_type;
            self.write_varint(key);
            value.write(self);
        }

        pub fn write_opt_field<V: Value>(
            &mut self,
            name: &str,
            field_number: u32,
            value: &Option<V>,
        ) {
            if let Some(value) = value {
                self.write_field(name, field_number, value);
            }
        }

        pub fn write_varint(&mut self, varint: u64) {
            leb128::write::unsigned(&mut self.buf, varint).unwrap();
        }
    }

    pub fn encode_message<M: Message>(message: &M) -> Vec<u8> {
        let mut w = MessageWriter::new();
        message.write(&mut w);
        w.buf
    }
}

#[derive(Clone, Copy)]
#[allow(dead_code)]
enum FwType {
    Application = 0,
    Softdevice = 1,
    Bootloader = 2,
    SoftdeviceAndBootloader = 3,
}

impl rohs::Value for FwType {
    const TYPE: WireType = WireType::Varint;

    fn write(&self, writer: &mut rohs::MessageWriter) {
        writer.write_varint(*self as _);
    }
}

#[derive(Clone, Copy)]
#[allow(dead_code)]
enum HashType {
    NoHash = 0,
    Crc = 1,
    Sha128 = 2,
    /// This is the *only* hash type the stock bootloader accepts.
    Sha256 = 3,
    Sha512 = 4,
}

impl rohs::Value for HashType {
    const TYPE: WireType = WireType::Varint;

    fn write(&self, writer: &mut rohs::MessageWriter) {
        writer.write_varint(*self as _);
    }
}

struct Hash<'a> {
    hash_type: HashType,
    hash: &'a [u8],
}

impl rohs::Message for Hash<'_> {
    fn write(&self, writer: &mut rohs::MessageWriter) {
        writer.write_field("hash_type", 1, &self.hash_type);
        writer.write_field("hash", 2, self.hash);
    }
}

struct InitCommand<'a> {
    // FIXME: expected structure is unclear here, all fields are optional in the upstream spec.
    // We just support the bare minimum.
    fw_version: u32,
    /// Marked as optional, but omitting it results in `InitCommandInvalid`.
    hw_version: u32,
    fw_type: FwType,
    sd_size: u32,
    bl_size: u32,
    /// Size of the flashed app image (total size of all data objects that follow).
    app_size: u32,
    /// Marked as optional in the proto file, but seems to be required.
    hash: Hash<'a>,
    is_debug: Option<bool>,
}

impl rohs::Message for InitCommand<'_> {
    fn write(&self, writer: &mut rohs::MessageWriter) {
        writer.write_field("fw_version", 1, &self.fw_version);
        writer.write_field("hw_version", 2, &self.hw_version);
        writer.write_field("type", 4, &self.fw_type);
        writer.write_field("sd_size", 5, &self.sd_size);
        writer.write_field("bl_size", 6, &self.bl_size);
        writer.write_field("app_size", 7, &self.app_size);
        writer.write_field("hash", 8, &self.hash);
        writer.write_opt_field("is_debug", 9, &self.is_debug);
    }
}

enum Command<'a> {
    InitCommand(InitCommand<'a>),
}

impl rohs::Message for Command<'_> {
    fn write(&self, writer: &mut rohs::MessageWriter) {
        match self {
            Command::InitCommand(cmd) => {
                writer.write_field("op_code", 1, &1);
                writer.write_field("init", 2, cmd);
            }
        }
    }
}

/// This is the outermost message, which will actually be sent to the bootloader.
enum Packet<'a> {
    Command(Command<'a>),
    // Missing: SignedCommand
}

impl rohs::Message for Packet<'_> {
    fn write(&self, writer: &mut rohs::MessageWriter) {
        match self {
            Packet::Command(cmd) => {
                writer.write_field("command", 1, cmd);
            }
        }
    }
}

pub fn build_init_packet(image: &[u8]) -> Vec<u8> {
    let mut hash = {
        let mut hasher = Sha256::new();
        hasher.update(image);
        hasher.finalize()
    };
    let hash = &mut *hash;
    // For some reason, Nordic insists on transmitting (and displaying) the hash in little-endian
    // byte order, unlike the entire rest of the industry.
    hash.reverse();

    log::debug!(
        "image size: {} Bytes ({} KiB)",
        image.len(),
        image.len() / 1024,
    );
    log::debug!(
        "image hash: {}",
        hash.iter()
            .map(|byte| format!("{:02x}", byte))
            .collect::<Vec<_>>()
            .join("")
    );

    let packet = Packet::Command(Command::InitCommand(InitCommand {
        fw_version: 0,
        // 52 is the default, the docs do not recommend using it, but it's unclear how to
        // accomplish that.
        hw_version: 52,
        fw_type: FwType::Application,
        sd_size: 0,
        bl_size: 0,
        app_size: image.len() as _,
        hash: Hash {
            hash_type: HashType::Sha256,
            hash,
        },
        is_debug: Some(false),
    }));

    rohs::encode_message(&packet)
}

#[cfg(test)]
mod tests {
    use super::*;
    use expect_test::{expect, Expect};

    fn test_message(debug_filename: &str, msg: impl rohs::Message, expect: Expect) {
        let bytes = rohs::encode_message(&msg);

        // Dump the encoded data to a file for easier inspection:
        // `protoc --decode_raw < tmp/file.dat`
        std::fs::create_dir_all("tmp").unwrap();
        std::fs::write(format!("tmp/{}.dat", debug_filename), &bytes).unwrap();

        let actual = format!("{:x?}", bytes);
        expect.assert_eq(&actual);
    }

    #[test]
    fn basic() {
        test_message(
            "test",
            Packet::Command(Command::InitCommand(InitCommand {
                fw_version: 0,
                hw_version: 52,
                fw_type: FwType::Application,
                sd_size: 0,
                bl_size: 0,
                app_size: 0x55,
                hash: Hash {
                    hash_type: HashType::Sha256,
                    hash: &[
                        0xae, 0x4b, 0x32, 0x80, 0xe5, 0x6e, 0x2f, 0xaf, 0x83, 0xf4, 0x14, 0xa6,
                        0xe3, 0xda, 0xbe, 0x9d, 0x5f, 0xbe, 0x18, 0x97, 0x65, 0x44, 0xc0, 0x5f,
                        0xed, 0x12, 0x1a, 0xcc, 0xb8, 0x5b, 0x53, 0xfc,
                    ],
                },
                is_debug: Some(true),
            })),
            expect![[
                r#"[a, 38, 8, 1, 12, 34, 8, 0, 10, 34, 20, 0, 28, 0, 30, 0, 38, 55, 42, 24, 8, 3, 12, 20, ae, 4b, 32, 80, e5, 6e, 2f, af, 83, f4, 14, a6, e3, da, be, 9d, 5f, be, 18, 97, 65, 44, c0, 5f, ed, 12, 1a, cc, b8, 5b, 53, fc, 48, 1]"#
            ]],
        );
    }
}
